File: 2023/MAR/19.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;


class Array
{
public:
    template <class T>
    static void input(vector<T> &A,int n)
    {
        for (int i = 0; i < n; i++)
        {
            scanf("%d ",&A[i]);
        }
    }

    template <class T>
    static void print(vector<T> &A)
    {
        for (int i = 0; i < A.size(); i++)
        {
            cout << A[i] << " ";
        }
        cout << endl;
    }
};


// } Driver Code Ends

class Solution {
  public:
    vector<int> getDistinctDifference(int n, vector<int> &a) {
        // code here
        vector<int> left(n,0);
        vector<int> right(n,0);
        set<int> s1;
        set<int> s2;
        
        s1.insert(a[0]);
        s2.insert(a[n-1]);
        for(int i=1 ; i<n ; i++)
        {
            int j = n-1-i;
            left[i] = s1.size();    s1.insert(a[i]);
            right[j] = s2.size();   s2.insert(a[j]);
        }
        
        vector<int> ans;
        for(int i=0 ; i<n ; i++)
        {
            ans.push_back(left[i]-right[i]);
        }
        return ans;
    }
};


//{ Driver Code Starts.

int main(){
    int t;
    scanf("%d ",&t);
    while(t--){
        
        int N;
        scanf("%d",&N);
        
        
        vector<int> A(N);
        Array::input(A,N);
        
        Solution obj;
        vector<int> res = obj.getDistinctDifference(N, A);
        
        Array::print(res);
        
    }
}

// } Driver Code Ends

File: 2023/MAR/19.py
from typing import List
class Solution:
    def getDistinctDifference(self, N : int, A : List[int]) -> List[int]:
        ans = list(A)
        left = list(A)
        right = list(A)
        left_set = set()
        right_set = set()
        for i in range(N):
            left[i] = len(left_set)
            left_set.add(A[i])
        for i in range(N-1,-1,-1):
            # print(i)
            right[i] = len(right_set)
            right_set.add(A[i])
            
        for i in range(N):
            ans[i] = left[i]-right[i]
        return ans
        
#{ 
 # Driver Code Starts
# class IntArray:

#     def __init__(self) -> None:
#         pass
    
if __name__=="__main__":
    t = int(input())
    for _ in range(t):
        
        N = int(input())
        
        A=[int(i) for i in input().split()]
        
        obj = Solution()
        res = obj.getDistinctDifference(N, A)
        
        print(*res)
        

# } Driver Code Ends
